src/main/scala/Index.scala

```diff
@@ -1,47 +1,32 @@
+import newmodel._
 
 
+case class Index(root: Pkg) {
+  val linear = loop(root)
 
-//trait Index {
-//  def getConversionsFor(id: ID): Seq[ImplicitConversions]
-//
-//  def getShadowedConversionsFor(id: ID): Seq[ImplicitConversions]
-//
-//  def getCompanionFor(id: ID): Option[ID]
-//
-//  def getLinearSupertypesBy(id: ID): Seq[ID]
-//
-//  def getKnownSubclasses(id: ID): Seq[ID]
-//
-//
-//}
-//
-//
-//object Index {
-//  def apply(docElements: Seq[DocElement]) = {
-//    val companionMap = buildCompanionMap(docElements)
-//
-//
-//    new Index {
-//      override def getKnownSubclasses(id: ID): Seq[ID] = ???
-//
-//      override def getShadowedConversionsFor(id: ID): Seq[ImplicitConversions] = ???
-//
-//      override def getLinearSupertypesBy(id: ID): Seq[ID] = ???
-//
-//      override def getCompanionFor(id: ID): Option[ID] = companionMap.get(id).map(_.id)
-//
-//      override def getConversionsFor(id: ID): Seq[ImplicitConversions] = ???
-//    }
-//
-//  }
-//
-//  def buildCompanionMap(docElements:Seq[DocElement]) ={
-//    docElements.map {
-//      case e: ClassDoc =>
-//        docElements.find({
-//          case ObjectDoc(`e`.name, _, _, _,_, `e`.file) => true
-//        }).map(e.id -> _)
-//      case _ => None
-//    }.filter(_.isDefined).map(_.get).toMap
-//  }
-//}
+  def classes = linear.collect { case e: Defn.Class => e }
+
+  def objects = linear.collect { case e: Defn.Object => e }
+
+  def traits = linear.collect { case e: Defn.Trait => e }
+
+  def defs = linear.collect { case e: Decl.Def => e }
+
+
+  private def mapper(stats: Seq[Stat]): Seq[Tree] = stats.map(loop(_)).flatten
+
+  private def loop(node: Tree): Seq[Tree] = {
+    val nodes: Seq[Tree] = node match {
+      case e: Defn.Class => mapper(e.templ.stats)
+      case e: Defn.Trait => mapper(e.templ.stats)
+      case e: Defn.Object => mapper(e.templ.stats)
+      case e: Pkg => mapper(e.stats)
+      case e: Decl.Def => Seq(e)
+      case _ => Seq()
+    }
+    node match {
+      case _: Defn.Class | _: Defn.Trait | _: Defn.Object | _: Pkg | _:Decl.Def => nodes ++ Seq(node)
+      case _ => nodes
+    }
+  }
```

Comment:
The meta to new-scaladoc translator should work very similar to this file... traversing the trees